waltz-tools/src/main/java/com/wepay/waltz/tools/performance/PerformanceCli.java

```diff
@@ -630,15 +669,18 @@ private PerformanceBase(String[] args) {
          */
         protected abstract class BaseCallbacks implements WaltzClientCallbacks {
 
-            protected long clientHighWaterMark;
+            protected Map<Integer, Long> partitionHighWaterMark = new HashMap<>();
 
-            private BaseCallbacks() {
-                clientHighWaterMark = -1;
+            private BaseCallbacks(Map<Integer, Long> partitionHighWaterMark) {
+                for (Map.Entry<Integer, Long> entry : partitionHighWaterMark.entrySet()) {
+                    this.partitionHighWaterMark.put(entry.getKey(), entry.getValue());
+                }
             }
 
             @Override
             public long getClientHighWaterMark(int partitionId) {
-                return clientHighWaterMark;
+                partitionHighWaterMark.putIfAbsent(partitionId, -1L);
+                return partitionHighWaterMark.get(partitionId);
```

Comment:
Why not just use `getOrDefault` that way you don't need `putIfAbsent` statement.

Reply:
Because of a check in _applyTransaction_ where transactionId is compared with high watermark of a transaction partition that is stored in _partitionHighWaterMarkMap_ and might be missing. This can be solved by using getOrDefault instead of get for every get statement.